File: javascript/src/problems101_200/word-break.js
class Problem {
  wordBreak(s, wordDict) {
    throw new Error(`Not implemented (${s}, ${wordDict}).`);
  }
}

class Default extends Problem {
  wordBreak(s, wordDict) {
    const dp = new Array(s.length + 1).fill(false);
    dp[0] = true;
    for (let right = 1; right < dp.length; right++) {
      for (let left = 0; left < right; left++) {
        if (!dp[left] || !wordDict.includes(s.substring(left, right))) {
          continue;
        }
        dp[right] = true;
        break;
      }
    }
    return dp[dp.length - 1];
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/word_break.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def word_break(self, s: str, word_dict: list[str]) -> bool:
        pass


class Default(Problem):
    @override
    def word_break(self, s: str, word_dict: list[str]) -> bool:
        dp: list[bool] = [False] * (len(s) + 1)
        dp[0] = True
        for right in range(1, len(dp)):
            for left in range(right):
                if not dp[left] or s[left:right] not in word_dict:
                    continue
                dp[right] = True
                break
        return dp[-1]


SOLUTIONS: list[Problem] = [Default()]
